fex2/preconfigured/experiments/common.sh

```diff
@@ -53,54 +78,44 @@ BUILD_DIR=${BUILD_DIR:-${BUILD_ROOT}/${NAME}}
 IFS=' ' read -r -a types <<<"$BUILD_TYPES"
 IFS=' ' read -r -a thread_counts <<<"$NUM_THREADS"
 
-is_benchmark_suite=0
-command=""
-execution_header=()
-input=""
-bin=""
-
-declare -a execution_header
+# These values are set in the run script of each experiment
+fex2_argument_is_benchmark_suite=false
+fex2_argument_benchmarks=""
+fex2_argument_command=""
 
 # shell output colors
-if [ $COLORED_LOGS -eq 1 ]; then
+if [ $COLORED_LOGS == true ]; then
     DEBUG='\033[94m\033[1m'
     WARNING='\033[93m\033[1m'
     ERROR='\033[91m\033[1m'
     ENDC='\033[0m'
-else
-    DEBUG=''
-    WARNING=''
-    ERROR=''
-    ENDC=''
 fi
 
 ##############################
 # Shortcuts and helpers
 ##############################
-function header_push() {
-    execution_header+=("$1")
-}
-
-function header_pop() {
-    local top=("${!execution_header[@]}")
-    local top_index=${#execution_header[@]}
-    if [ "$top_index" -le 0 ]; then
-        echo "Current Header: " "${execution_header[@]}"
-        error_exit "header_pop: unbalanced pop from the header" 1
-    fi
-    unset 'execution_header[${top[@]: -1}]'
```

Comment:
I find the code quite confusing. Is the `execution_header` in the `build` / `single_execution` function not as good as your header? :thinking:

Reply:
The goal was to have easy variability of the experiment setups. Below, there is a line:
```
for_types for_benchmarks for_thread_counts for_iterations single_execution
```
With this header stack, the user can easily change the order of iteration, remove some of the loops, or add new. All with a single line of code modified, and in a single function. To do so, the user would need to only redefine the `execute_experiment` function.

On the other hand, it you hardcode the header, then it will be inconsistent with the updated experiment and would need to be modified as well.